README.rst
===============================
django-rsvp: A simple RSVP app.
===============================

``django-rsvp`` is a simple RSVP application for use with Django. The intent
is to be able to basic events, send guests invite e-mails and collect their
response if they will be attending or not.


Requirements
============

``django-rsvp`` requires:

* Python 2.3+
* Django 1.0+

The only potential dependency within Django is that ``django.contrib.sites``
is in ``INSTALLED_APPS`` in order to make the included e-mail template work.


Installation
============

#. Either copy/symlink the ``rsvp`` app into your project or place it
   somewhere on your ``PYTHONPATH``.
#. Add ``rvsp`` app to your ``INSTALLED_APPS``.
#. Set the ``RSVP_FROM_EMAIL`` setting to an e-mail address you'd like
   invites to be sent from.
#. Run ``./manage.py syncdb``.
#. Add ``(r'^rsvp/', include('rsvpproject.rsvp.urls')),`` to your
   ``urls.py``.


CHANGELOG.txt
v1.1
====

* Refactored views for better reusability.
* Refactored models for better reusability.
* Fixed GuestInlines in admin.


v1.0
====

* Initial release.

rsvp/tests.py
"""
>>> from django.core.management import call_command
>>> call_command('loaddata', 'rsvp_testdata.yaml') #doctest: +ELLIPSIS
Installing yaml fixture 'rsvp_testdata' ...
Installed 4 object(s) from 1 fixture(s)

>>> from django.test import Client
>>> c = Client()

>>> from django.core import mail
>>> from django.core.urlresolvers import reverse
>>> from rsvp.models import Event, Guest

# Test sending the e-mails.
>>> mail.outbox
[]
>>> event = Event.objects.get(pk=1)
>>> event
<Event: A Test Event>
>>> event.send_guest_emails()
2
>>> len(mail.outbox)
2
>>> mail.outbox[0].to
[u'guest2@example.com']
>>> mail.outbox[0].subject
u'Come to my test event!'
>>> mail.outbox[0].body
u'We will have fun.\\n\\nTo RVSP to this invite, please visit http://example.com/rsvp/event/a-test-event/.'

# Re-empty the bin.
>>> mail.outbox = []

>>> r = c.get(reverse('rsvp_event_view', args=['a-test-event']))
>>> r.status_code
200
>>> r.context[-1]['event']
<Event: A Test Event>
>>> r.context[-1]['event'].guests_attending()
[<Guest: A Test Event - guest1@example.com - yes>]
>>> r.context[-1]['event'].guests_no_rsvp()
[<Guest: A Test Event - guest2@example.com - no_rsvp>, <Guest: A Test Event - guest3@example.com - no_rsvp>]
>>> type(r.context[-1]['form'])
<class 'rsvp.forms.RSVPForm'>

>>> r = c.post(reverse('rsvp_event_view', args=['a-test-event']), {
...     'email': 'guest99@example.com',
...     'name': 'Guest #99',
...     'attending': 'yes',
...     'number_of_guests': '0',
...     'comment': '',
... })
>>> r.status_code
200
>>> r.context[-1]['form'].errors
{'email': [u'That e-mail is not on the guest list.']}

>>> r = c.post(reverse('rsvp_event_view', args=['a-test-event']), {
...     'email': 'guest1@example.com',
...     'name': 'Guest #1',
...     'attending': 'yes',
...     'number_of_guests': '0',
...     'comment': '',
... })
>>> r.status_code
200
>>> r.context[-1]['form'].errors
{'email': [u'You have already provided RSVP information.']}

>>> r = c.post(reverse('rsvp_event_view', args=['a-test-event']), {
...     'email': 'guest2@example.com',
...     'name': 'Mr. Guest #2',
...     'attending': 'yes',
...     'number_of_guests': '-1',
...     'comment': '',
... })
>>> r.status_code
200
>>> r.context[-1]['form'].errors
{'number_of_guests': [u"The number of guests you're bringing can not be negative."]}

>>> r = c.post(reverse('rsvp_event_view', args=['a-test-event']), {
...     'email': 'guest2@example.com',
...     'name': 'Mr. Guest #2',
...     'attending': 'yes',
...     'number_of_guests': '2',
...     'comment': 'Happy to come and bringing a dish!',
... })
>>> r.status_code
302
>>> r['location']
'http://testserver/rsvp/event/a-test-event/thanks/2/'

>>> r = c.get(reverse('rsvp_event_view', args=['a-test-event']))
>>> r.status_code
200
>>> r.context[-1]['event']
<Event: A Test Event>
>>> r.context[-1]['event'].guests_attending()
[<Guest: A Test Event - guest1@example.com - yes>, <Guest: A Test Event - guest2@example.com - yes>]
>>> r.context[-1]['event'].guests_no_rsvp()
[<Guest: A Test Event - guest3@example.com - no_rsvp>]

>>> r = c.post(reverse('rsvp_event_view', args=['a-test-event']), {
...     'email': 'guest2@example.com',
...     'name': 'Mr. Guest #2',
...     'attending': 'yes',
...     'number_of_guests': '2',
...     'comment': 'Happy to come and bringing a dish!',
... })
>>> r.status_code
200
>>> r.context[-1]['form'].errors
{'email': [u'You have already provided RSVP information.']}
"""


rsvp/admin.py
from django.contrib import admin
from rsvp.models import Event, Guest


class GuestInline(admin.TabularInline):
    model = Guest
    extra = 3
    fields = ('email', 'name', 'attending_status', 'number_of_guests')


class EventAdmin(admin.ModelAdmin):
    date_hiearchy = 'date_of_event'
    fieldsets = (
        (None, {
            'fields': ('title', 'slug', 'description', 'date_of_event'),
        }),
        ('Email', {
            'fields': ('email_subject', 'email_message'),
        }),
        ('Event Details', {
            'fields': ('hosted_by', 'street_address', 'city', 'state', 'zip_code', 'telephone'),
            'classes': ('collapse',)
        })
    )
    inlines = [GuestInline]
    list_display = ('title', 'date_of_event')
    prepopulated_fields = {'slug': ('title',)}
    search_fields = ('title', 'description', 'hosted_by')


class GuestAdmin(admin.ModelAdmin):
    fields = ('event', 'email', 'name', 'attending_status', 'number_of_guests', 'comment')
    list_display = ('email', 'name', 'attending_status', 'number_of_guests')
    list_filter = ('attending_status',)
    search_fields = ('email', 'name')


admin.site.register(Event, EventAdmin)
admin.site.register(Guest, GuestAdmin)


rsvp/views.py
import datetime
from django.shortcuts import render_to_response, get_object_or_404
from django.http import HttpResponseRedirect, Http404
from django.core.exceptions import ObjectDoesNotExist
from django.core.urlresolvers import reverse
from django.template import RequestContext
from rsvp.models import Event, Guest
from rsvp.forms import RSVPForm


def event_view(request, slug, model_class=Event, form_class=RSVPForm, template_name='rsvp/event_view.html'):
    event = get_object_or_404(model_class, slug=slug)
    
    if request.POST:
        form = form_class(request.POST)
        
        if form.is_valid():
            guest = form.save()
            return HttpResponseRedirect(reverse('rsvp_event_thanks', kwargs={'slug': slug, 'guest_id': guest.id}))
    else:
        form = form_class()
    
    return render_to_response(template_name, {
        'event': event,
        'form': form,
    }, context_instance=RequestContext(request))


def event_thanks(request, slug, guest_id, model_class=Event, template_name='rsvp/event_thanks.html'):
    event = get_object_or_404(model_class, slug=slug)
    
    try:
        guest = event.guests.get(pk=guest_id)
    except ObjectDoesNotExist:
        raise Http404
    
    return render_to_response(template_name, {
        'event': event,
        'guest': guest,
    }, context_instance=RequestContext(request))

rsvp/models.py
import datetime
from django.db import models
from django.db.models import permalink
from django.core.mail import send_mass_mail
from django.template import loader, Context
from django.conf import settings
from django.contrib.sites.models import Site


ATTENDING_CHOICES = (
    ('yes', 'Yes'),
    ('no', 'No'),
    ('maybe', 'Maybe'),
    ('no_rsvp', 'Hasn\'t RSVPed yet')
)


class Event(models.Model):
    title = models.CharField(max_length=255)
    slug = models.SlugField()
    description = models.TextField()
    date_of_event = models.DateTimeField()
    verification_code = models.CharField(max_length=32, blank=True, default='', help_text='Present for future extension/guest verification.')
    email_subject = models.CharField(max_length=255, help_text='The subject line for the e-mail sent out to guests.')
    email_message = models.TextField(help_text='The body of the e-mail sent out to guests.')
    hosted_by = models.CharField(max_length=255, help_text='The name of the person/organization hosting the event.', blank=True, default='')
    street_address = models.CharField(max_length=255, help_text='The street address where the event is being held.', blank=True, default='')
    city = models.CharField(max_length=64, help_text='The city where the event is being held.', blank=True, default='')
    state = models.CharField(max_length=64, help_text='The state where the event is being held.', blank=True, default='')
    zip_code = models.CharField(max_length=10, help_text='The zip code where the event is being held.', blank=True, default='')
    telephone = models.CharField(max_length=20, blank=True, default='')
    created = models.DateTimeField(default=datetime.datetime.now)
    updated = models.DateTimeField(blank=True, null=True)
    
    def __unicode__(self):
        return self.title
    
    def save(self):
        self.updated = datetime.datetime.now()
        super(Event, self).save()
    
    def get_absolute_url(self):
        return ('rsvp_event_view', [self.slug])
    get_absolute_url = permalink(get_absolute_url)
    
    def guests_attending(self):
        return self.guests.filter(attending_status='yes')
    
    def guests_not_attending(self):
        return self.guests.filter(attending_status='no')
    
    def guests_may_attend(self):
        return self.guests.filter(attending_status='maybe')
    
    def guests_no_rsvp(self):
        return self.guests.filter(attending_status='no_rsvp')
    
    def send_guest_emails(self):
        """
        Sends an invite e-mail to all guest who have not RSVPed.
        
        Requires settings RSVP_FROM_EMAIL in your settings file. Returns a 
        count of the number of guests e-mailed.
        """
        mass_mail_data = []
        from_email = getattr(settings, 'RSVP_FROM_EMAIL', '')
        
        for guest in self.guests_no_rsvp():
            t = loader.get_template('rsvp/event_email.txt')
            c = Context({
                'event': self,
                'site': Site.objects.get_current(),
            })
            message = t.render(c)
            mass_mail_data.append([self.email_subject, message, from_email, [guest.email]])
        
        send_mass_mail(mass_mail_data, fail_silently=True)
        return self.guests_no_rsvp().count()


class Guest(models.Model):
    event = models.ForeignKey(Event, related_name='guests')
    email = models.EmailField()
    name = models.CharField(max_length=128, blank=True, default='')
    attending_status = models.CharField(max_length=32, choices=ATTENDING_CHOICES, default='no_rsvp')
    number_of_guests = models.SmallIntegerField(default=0)
    comment = models.CharField(max_length=255, blank=True, default='')
    created = models.DateTimeField(default=datetime.datetime.now)
    updated = models.DateTimeField(blank=True, null=True)
    
    def __unicode__(self):
        return u"%s - %s - %s" % (self.event.title, self.email, self.attending_status)
    
    def save(self):
        self.updated = datetime.datetime.now()
        super(Guest, self).save()


rsvp/forms.py
from django import forms
from django.core.exceptions import ObjectDoesNotExist
from rsvp.models import ATTENDING_CHOICES, Guest


VISIBLE_ATTENDING_CHOICES = [choice for choice in ATTENDING_CHOICES if choice[0] != 'no_rsvp']


class RSVPForm(forms.Form):
    email = forms.EmailField()
    name = forms.CharField(max_length=128)
    attending = forms.ChoiceField(choices=VISIBLE_ATTENDING_CHOICES, initial='yes', widget=forms.RadioSelect)
    number_of_guests = forms.IntegerField(initial=0)
    comment = forms.CharField(max_length=255, required=False, widget=forms.Textarea)
    
    def __init__(self, *args, **kwargs):
        if 'guest_class' in kwargs:
            self.guest_class = kwargs['guest_class']
            del(kwargs['guest_class'])
        else:
            self.guest_class = Guest
        super(RSVPForm, self).__init__(*args, **kwargs)
    
    def clean_email(self):
        try:
            guest = self.guest_class._default_manager.get(email=self.cleaned_data['email'])
        except ObjectDoesNotExist:
            raise forms.ValidationError, 'That e-mail is not on the guest list.'
        
        if hasattr(guest, 'attending_status') and guest.attending_status != 'no_rsvp':
            raise forms.ValidationError, 'You have already provided RSVP information.'
        
        return self.cleaned_data['email']
    
    def clean_number_of_guests(self):
        if self.cleaned_data['number_of_guests'] < 0:
            raise forms.ValidationError, "The number of guests you're bringing can not be negative."
        return self.cleaned_data['number_of_guests']
        
    def save(self):
        guest = self.guest_class._default_manager.get(email=self.cleaned_data['email'])
        
        if self.cleaned_data['name']:
            guest.name = self.cleaned_data['name']
        
        guest.attending_status = self.cleaned_data['attending']
        guest.number_of_guests = self.cleaned_data['number_of_guests']
        guest.comment = self.cleaned_data['comment']
        guest.save()
        return guest
